slice-core-api/src/main/java/com/cognifide/slice/api/injector/InjectorRunner.java

```diff
@@ -87,11 +91,32 @@ public void installModules(final List<? extends Module> newModules) {
 		modules.addAll(newModules);
 	}
 
-	public void start() {
-		InjectorConfig config = new InjectorConfig(this);
+	public void start() throws BundleException {
+		final InjectorConfig config = new InjectorConfig(this);
+
+		config.setListener(new InjectorCreationFailListener() {
+			@Override public void creationFailed() {
+				try {
+					injectorCreationSuccess = false;
+					started = false;
+					bundleContext.getBundle().stop();
+					//bundleContext.getBundle().start();//this would leave bundle in RESOLVED state instead of INSTALLED
+				} catch (BundleException e) {
+					e.printStackTrace();
+				}
+			}
+		});
+
 		Dictionary<String, Object> properties = new Hashtable<String, Object>();
-		bundleContext.registerService(InjectorConfig.class.getName(), config, properties);
-		started = true;
+		registration = bundleContext.registerService(InjectorConfig.class.getName(), config, properties);
+
+		if (injectorCreationSuccess) {
```

Comment:
In this condition we basically rely on the race. We hope that the registered config will be intercepted by the InjectorHierarchy class, that this class will try to create an injector and fail before the current thread gets to this place. We shouldn't do it. The method should end at `bundleContext.registerService()` and after that we should handle the whole injector creation fail logic in the `InjectorCreationFailListener`.

Please notice, that the injector may be created asynchronously. Eg. you have injector `myApp` with parent `zg`. Slice won't try to create the `myApp` until the `zg` is available.